ldb_reader.go

```diff
@@ -490,3 +570,110 @@ func (reader *LDBReader) getGetRowByKeyStmt(ctx context.Context, pk schema.Prima
 
 	return stmt, err
 }
+
+func (reader *LDBReader) watchForLDBs(ctx context.Context, dirPath string, last int64) {
+	// TODO: consider using a fs notification library instead of polling.
+	ticker := time.NewTicker(time.Second)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			fsLast, err := lookupLastLDBSync(dirPath)
+			if err != nil {
+				events.Log("failed checking for last LDB sync: %{error}+v", err)
+				errs.Incr("check-last-ldb-sync")
+				continue
+			}
+
+			// Only swap LDBs if this LDB is newer than our newest LDB.
+			if fsLast <= last {
+				continue
+			}
+			events.Log("found new LDB (%d > %d), switching...", fsLast, last)
+			last = fsLast
+
+			err = reader.switchToLDB(dirPath, last)
+			if err != nil {
+				events.Log("failed switching to new LDB: %{error}+v", err)
+				errs.Incr("switch-ldb")
+			}
+		}
+	}
+}
+
+func (reader *LDBReader) switchToLDB(dirPath string, timestamp int64) error {
+	fullPath := filepath.Join(dirPath, fmt.Sprintf("%013d", timestamp), ldb.DefaultLDBFilename)
+
+	db, err := newLDB(fullPath)
+	if err != nil {
+		return errors.Wrap(err, "new ldb")
+	}
+
+	reader.mu.Lock()
+	defer reader.mu.Unlock()
+
+	if err = reader.closeDB(); err != nil {
+		return errors.Wrap(err, "closing db")
+	}
+
+	reader.Db = db
+
+	return nil
+}
+
+func lookupLastLDBSync(dirPath string) (int64, error) {
+	// Loop through the files in the `dirPath` and look for the ldb.db with the
+	// highest associated timestamp. Return that.
+	var lastSync int64
+	err := filepath.Walk(dirPath, func(filePath string, info os.FileInfo, _ error) error {
+		// Ignore directories:
+		if info.IsDir() {
+			return nil
+		}
+
+		// We only care about the timestamps associated with `<path>/<timestamp>/ldb.db` files.
```

Comment:
think it would be a good idea to use <path>/versioned/<timestamp> to help future generations understand what these other ldbs are... :)